[FSDP2] Removed logic to save and remove pre-backward hook handles

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125269
* #125191
* #125190

1. This PR removes the logic for saving and removing the pre-backward hook handles (which is registered via `register_multi_grad_hook(mode="any")`).
2. This PR removes the logic for _trying_ to guard against mistargeted prefetches that relies on querying if the engine will execute the module output tensors' `grad_fn`s. (See https://github.com/pytorch/pytorch/pull/118118 for original motivation.)

For 1, the logic was error prone since it relied on `set_is_last_backward(False)` being set correctly or else pre-backward hooks could be de-registered too early. We would prefer to match the hook lifetimes with that of the autograd graph. This solves a bug with a 1f1b interleaved schedule.

If we directly remove the manual saving/removing hook handle logic, then we have a ref cycle where the tensors' `grad_fn`s are passed to the hook function. We decide to simply remove this `grad_fn` logic since (1) it cannot perfectly prevent mistargeted prefetches and (2) it introduces undesired complexity. In the future, we may prefer a different mechanism to override the prefetching for more complex/dynamic use cases.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
[FSDP2] Removed logic to save and remove pre-backward hook handles

[ghstack-poisoned]

awgu commented:
To-do: there is a ref cycle because the hook function closes over the `grad_fn`s

Commit:
Update on "[FSDP2] Removed logic to save and remove pre-backward hook handles"


We should not necessarily remove the pre-backward hook at the end of backward in the case that there may be following microbatch backwards. Anyway, there is no motivation to keep a reference to the hook handle, as its lifetime matches that of the autograd graph, which is what we want. Let us remove this logic to simplify.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Removed logic to save and remove pre-backward hook handles"


1. This PR removes the logic for saving and removing the pre-backward hook handles (which is registered via `register_multi_grad_hook(mode="any")`).
2. This PR removes the logic for _trying_ to guard against mistargeted prefetches that relies on querying if the engine will execute the module output tensors' `grad_fn`s. (See https://github.com/pytorch/pytorch/pull/118118 for original motivation.)

For 1, the logic was error prone since it relied on `set_is_last_backward(False)` being set correctly or else pre-backward hooks could be de-registered too early. We would prefer to match the hook lifetimes with that of the autograd graph. This solves a bug with a 1f1b interleaved schedule.

If we directly remove the manual saving/removing hook handle logic, then we have a ref cycle where the tensors' `grad_fn`s are passed to the hook function. We decide to simply remove this `grad_fn` logic since (1) it cannot perfectly prevent mistargeted prefetches and (2) it introduces undesired complexity. In the future, we may prefer a different mechanism to override the prefetching for more complex/dynamic use cases.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy reviewed (approved):
I am totally good with falling back to reverse order prefetching

remotely related to this PR: for mismatch warning (maybe also applies without this PR), I remember we had warnings in FSDP1. Is the mismatching warning possible in FSDP2?

awgu commented:
@weifengpy Hmm, I do not remember there being warnings for mistargeted prefetches in backward. Looking in `_runtime_utils.py` and `_exec_order_utils.py`, I do not see it either.

There is some warning if the execution order changes in a later iteration (only in `TORCH_DISTRIBUTED_DEBUG=DETAIL`):
https://github.com/pytorch/pytorch/blob/1bcbc9158ff7093a7d8ff3304f70dcc5e6294326/torch/distributed/fsdp/_exec_order_utils.py#L298-L302

We could possibly add some warning logic, but I think it might be simpler to add something like `TORCH_LOGS` integration where we log `unshard`/`reshard` ops including if the `unshard` was prefetched (among other things like pre/post-forward/backward hooks running for example).

Commit:
Update on "[FSDP2] Removed logic to save and remove pre-backward hook handles"


1. This PR removes the logic for saving and removing the pre-backward hook handles (which is registered via `register_multi_grad_hook(mode="any")`).
2. This PR removes the logic for _trying_ to guard against mistargeted prefetches that relies on querying if the engine will execute the module output tensors' `grad_fn`s. (See https://github.com/pytorch/pytorch/pull/118118 for original motivation.)

For 1, the logic was error prone since it relied on `set_is_last_backward(False)` being set correctly or else pre-backward hooks could be de-registered too early. We would prefer to match the hook lifetimes with that of the autograd graph. This solves a bug with a 1f1b interleaved schedule.

If we directly remove the manual saving/removing hook handle logic, then we have a ref cycle where the tensors' `grad_fn`s are passed to the hook function. We decide to simply remove this `grad_fn` logic since (1) it cannot perfectly prevent mistargeted prefetches and (2) it introduces undesired complexity. In the future, we may prefer a different mechanism to override the prefetching for more complex/dynamic use cases.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy commented:
> @weifengpy Hmm, I do not remember there being warnings for mistargeted prefetches in backward. Looking in `_runtime_utils.py` and `_exec_order_utils.py`, I do not see it either.
> 
> There is some warning if the execution order changes in a later iteration (only in `TORCH_DISTRIBUTED_DEBUG=DETAIL`):
> 
> https://github.com/pytorch/pytorch/blob/1bcbc9158ff7093a7d8ff3304f70dcc5e6294326/torch/distributed/fsdp/_exec_order_utils.py#L298-L302
> 
> We could possibly add some warning logic, but I think it might be simpler to add something like `TORCH_LOGS` integration where we log `unshard`/`reshard` ops including if the `unshard` was prefetched (among other things like pre/post-forward/backward hooks running for example).

you're right about FSDP1. There is only mismatch in forward orders. Got you about the TORCH_LOGS

detour again: I guess we only do implicit prefetching in FSDP2? It means no such thing of forward mismtach?

awgu commented:
>  I guess we only do implicit prefetching in FSDP2? It means no such thing of forward mismtach?

Yep, there should be no forward mismatch in FSDP2 (for now while we do not have explicit forward prefetching).

awgu commented:
@pytorchbot merge
